src/Features/CSharp/Portable/ForeachToFor/CSharpForeachToForCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,160 @@
+                                default,
+                                SyntaxFactory.EqualsValueClause(
+                                    foreachInfo.RequireExplicitCast ?
+                                        SyntaxFactory.CastExpression(
+                                            SyntaxFactory.ParseTypeName(foreachInfo.ExplicitCastInterface).WithAdditionalAnnotations(Simplifier.Annotation),
+                                            SyntaxFactory.ParenthesizedExpression(foreachCollectionExpression)) : 
+                                        foreachCollectionExpression)))));
+
+                collectionStatement =
+                    collectionStatement.ReplaceToken(collectionStatement.GetFirstToken(), collectionStatement.GetFirstToken().WithLeadingTrivia(foreachStatement.ForEachKeyword.LeadingTrivia));
+
+                editor.InsertBefore(foreachStatement, collectionStatement);
+            }
+
+            // create new index varialbe name
+            var indexString = CreateUniqueMethodName(model, foreachStatement.Statement, "i");
+
+            // put variable statement in body
+            var bodyStatement = GetForLoopBody(foreachInfo, parseOption, collectionVariableName, indexString);
+
+            // create for statement from foreach statement
+            var forStatement = SyntaxFactory.ForStatement(
+                SyntaxFactory.VariableDeclaration(SyntaxFactory.IdentifierName("var"),
+                    SyntaxFactory.SingletonSeparatedList(
+                        SyntaxFactory.VariableDeclarator(
+                        SyntaxFactory.Identifier(indexString).WithAdditionalAnnotations(RenameAnnotation.Create()),
+                        default,
+                        SyntaxFactory.EqualsValueClause(
+                            SyntaxFactory.LiteralExpression(SyntaxKind.NumericLiteralExpression, SyntaxFactory.Literal(0)))))),
+                SyntaxFactory.SeparatedList<ExpressionSyntax>(),
+                SyntaxFactory.BinaryExpression(SyntaxKind.LessThanExpression, SyntaxFactory.IdentifierName(indexString), SyntaxFactory.ParseExpression($"{collectionVariableName}.{foreachInfo.CountName}")),
+                SyntaxFactory.SingletonSeparatedList<ExpressionSyntax>(SyntaxFactory.PostfixUnaryExpression(SyntaxKind.PostIncrementExpression, SyntaxFactory.IdentifierName(indexString))),
+                bodyStatement);
+
+            // let leading and trailing trivia set
+            if (!foreachInfo.RequireCollectionStatement)
+            {
+                forStatement = forStatement.ReplaceToken(forStatement.ForKeyword, forStatement.ForKeyword.WithLeadingTrivia(foreachStatement.ForEachKeyword.LeadingTrivia));
+            }
+
+            forStatement = forStatement.ReplaceToken(forStatement.CloseParenToken, forStatement.CloseParenToken.WithTrailingTrivia(foreachStatement.CloseParenToken.TrailingTrivia));
```

Comment:
this seems super weird.  why not just create the for-statment, passing in the foreach statement closing paren?

Reply:
because I dont want to give in all tokens when I use SyntaxFactory. and I don't see why it is weird. you get same result. no reason to harder way when there is other simple way to do so.

Reply:
it's weird because you constructed a node, only to mutate that node.  And you mutated it solely to give it a token you already have.  The general pattern we use here is to just construct the node with the tokens it needs.  less mutation is easier to read and understand.

--

Also, if you wanted to mutate, i don't see why you didn't just write:

```c#
forStatement = forStatement.ReplaceToken(forStatement.CloseParenToken, foreachStatement.CloseParenToken);
```

That seems much simpler and cleaner.

Finally, i don't understand why we're keeping the close paren, but not the open paren.  In our features that mutate/rewrite existing code, we try to keep as many of the original tokens in the final form so taht we can preserve as much user intent and trivia as possible.